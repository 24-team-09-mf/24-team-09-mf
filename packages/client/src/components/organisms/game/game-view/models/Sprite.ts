import { GameModel, SpriteModel } from '@/components/organisms/game/game-view/game-view.types'

export class Sprite {
  position = {
    x: 0,
    y: 0,
  }

  dimensions = {
    width: 0,
    height: 0,
  }

  cropbox?= {
    width: 0,
    height: 0,
  }

  model: GameModel = null
  color = ''
  imageSrc?= ''


  constructor({ position, model, dimensions, color, cropbox, imageSrc }: SpriteModel) {
    this.position = position;
    this.model = model;
    this.dimensions = dimensions
    this.cropbox = cropbox
    this.color = color || 'red'
    this.imageSrc = imageSrc || undefined
  }

  draw() {
    if (this.model) {
      if (this.imageSrc) {
        console.log(this.imageSrc);
        const image = new Image();
        image.src = this.imageSrc;
        image.onload = () => {
          if (this.cropbox) {
            this.model?.drawImage(
              image,
              0,
              0,
              this.cropbox.width,
              this.cropbox.height,
              this.position.x,
              this.position.y,
              this.dimensions.width,
              this.dimensions.height
            );
          }
        }
      }
      else {
        this.model.fillStyle = this.color;
        this.model.fillRect(
          this.position.x,
          this.position.y,
          this.dimensions.width,
          this.dimensions.height
        )
      }
    }
  }
}
