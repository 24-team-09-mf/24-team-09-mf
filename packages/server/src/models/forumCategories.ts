import {
  DataType,
  Model,
  Table,
  Column,
  AllowNull,
  Length,
  HasMany,
} from 'sequelize-typescript'
import { TopicsModel } from './forumTopics'

@Table({ modelName: 'categories' })
export class CategoriesModel extends Model {
  @AllowNull(false)
  @Length({ min: 3, max: 255 })
  @Column(DataType.STRING)
  title: string

  @AllowNull(true)
  @Column(DataType.STRING)
  description: string

  @HasMany(() => TopicsModel, 'parent_id')
  topics: Array<TopicsModel>
}
