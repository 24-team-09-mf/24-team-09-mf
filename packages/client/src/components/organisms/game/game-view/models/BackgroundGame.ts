import { Sprite } from '@/components/organisms/game/game-view/models/Sprite'
import { SpriteModel } from '@/components/organisms/game/game-view/game-view.types'

export class BackgroundGame extends Sprite {
  constructor(props: SpriteModel) {
    super(props)
  }

  draw() {
    if (this.model) {
      this.model.drawImage(
        this.image,
        this.position.x,
        this.position.y,
        this.dimensions.width,
        this.dimensions.height
      )
    }
  }
}
