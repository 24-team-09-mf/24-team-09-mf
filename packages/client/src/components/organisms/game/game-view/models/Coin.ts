import { Sprite } from '@/components/organisms/game/game-view/models/Sprite'
import { SpriteModel } from '@/components/organisms/game/game-view/game-view.types'

export class Coin extends Sprite {
  shouldDraw = true
  constructor(props: SpriteModel) {
    super(props)
  }

  getCoin() {
    this.shouldDraw = false
    console.log('get coin, you should update score on one')
  }
  draw() {
    if (this.shouldDraw) {
      super.draw()
    }
  }
}
