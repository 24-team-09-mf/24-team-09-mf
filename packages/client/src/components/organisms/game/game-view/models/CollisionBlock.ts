import { Sprite } from '@/components/organisms/game/game-view/models/Sprite'
import { SpriteModel } from '@/components/organisms/game/game-view/game-view.types'

export class CollisionBlock extends Sprite {
  constructor(props: SpriteModel) {
    super(props)
  }

  draw() {
    super.draw()
  }
}
