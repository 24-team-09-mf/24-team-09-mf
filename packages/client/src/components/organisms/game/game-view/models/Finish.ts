import { Sprite } from '@/components/organisms/game/game-view/models/Sprite'
import { SpriteModel } from '@/components/organisms/game/game-view/game-view.types'

export class Finish extends Sprite {
  finished = false
  onGameOver: (outcome: string) => void

  constructor(props: SpriteModel & { onGameOver: (outcome: string) => void }) {
    super(props)
    this.onGameOver = props.onGameOver
  }

  getFinish(callback: () => void) {
    if (!this.finished) {
      callback()
      this.finished = true
      this.onGameOver('win')

    }
  }
}
