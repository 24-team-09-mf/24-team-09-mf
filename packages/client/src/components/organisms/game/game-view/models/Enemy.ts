import { Sprite } from '@/components/organisms/game/game-view/models/Sprite'
import { CollisionBlock } from '@/components/organisms/game/game-view/models/CollisionBlock'
import { SpriteModel } from '@/components/organisms/game/game-view/game-view.types'

type Props = SpriteModel & {
  collisionBlocks: CollisionBlock[]
  startVelocity: number
  imageSrc?: string
}
export class Enemy extends Sprite {
  collisionBlocks: CollisionBlock[] = []
  velocity = {
    x: 0,
  }
  shouldDraw = true
  constructor({
    collisionBlocks,
    position,
    startVelocity,
    imageSrc,
    frameRate,
    animations,
    ...props
  }: Props) {
    super({ ...props, position, imageSrc, frameRate, animations })
    this.collisionBlocks = collisionBlocks
    this.position = position
    this.velocity.x = startVelocity
  }

  destroyEnemy() {
    this.shouldDraw = false
    console.log('destroy enemy')
  }

  update() {
    this.position.x += this.velocity.x
    this.checkHorizontalCollision()
    this.draw()
  }

  switchSprite(name: string) {

    const animation = this.animations?.[name]
    if (this.image === animation?.image) return

    this.currentFrame = 0

    if (animation?.image) {
      this.image = animation.image
      this.frameRate = animation.frameRate || 1
      this.frameBuffer = animation.frameBuffer || 2
    }
  }

  checkHorizontalCollision() {
    for (let i = 0; i < this.collisionBlocks.length; i++) {
      const collisionBlock = this.collisionBlocks[i]
      if (this.checkCollision(collisionBlock)) {
        this.velocity.x = this.velocity.x * -1
        const animationName = this.velocity.x > 0 ? 'runRight' : 'runLeft'
        this.switchSprite(animationName)
        break
      }
    }
  }

  checkCollision(collisionBlock: CollisionBlock) {
    return (
      this.position.x <=
      collisionBlock.position.x + collisionBlock.dimensions.width &&
      this.position.x + this.dimensions.width >= collisionBlock.position.x &&
      this.position.y === collisionBlock.position.y
    )
  }

  draw() {
    if (this.shouldDraw) super.draw()
  }
}
