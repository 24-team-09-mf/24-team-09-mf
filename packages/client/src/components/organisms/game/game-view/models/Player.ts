import { SpriteModel } from '@/components/organisms/game/game-view/game-view.types'
import { Sprite } from '@/components/organisms/game/game-view/models/Sprite'
import { HEIGHT_VIEW } from '@/components/organisms/game/game.constants'
import { CollisionBlock } from '@/components/organisms/game/game-view/models/CollisionBlock'
import { Coin } from '@/components/organisms/game/game-view/models/Coin'
import { Enemy } from '@/components/organisms/game/game-view/models/Enemy'

type Props = SpriteModel & {
  collisionBlocks: CollisionBlock[]
  coins: Coin[]
  enemies: Enemy[]
  onGameOver: () => void
  color?: string
  imageSrc?: string
}
export class Player extends Sprite {

  velocity = {
    x: 0,
    y: 0,
  }

  lastDirection = ''
  gravity = 0.6
  collisionBlocks: CollisionBlock[] = []
  coins: Coin[] = []
  enemies: Enemy[] = []

  constructor({
    position,
    model,
    collisionBlocks,
    coins,
    enemies,
    onGameOver,
    imageSrc, color, frameRate, animations
  }: Props) {
    super({ position, model, imageSrc, color, frameRate, animations })
    this.position = position
    this.model = model
    this.collisionBlocks = collisionBlocks
    this.coins = coins
    this.enemies = enemies
    this.gameOver = onGameOver
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  gameOver() { }

  update() {
    this.position.x += this.velocity.x
    this.checkHorizontalCollision()
    this.velocity.y += this.gravity
    this.position.y += this.velocity.y
    this.checkVerticalCollision()

    this.checkCoinCollision()
    this.checkEnemiesCollision()
    if (
      this.position.y + this.dimensions.height + this.velocity.y >
      HEIGHT_VIEW
    ) {
      this.gameOver()
    }
    this.draw()
  }

  switchSprite(name: string) {

    const animation = this.animations?.[name]
    if (this.image === animation?.image) return

    this.currentFrame = 0

    if (animation?.image) {
      this.image = animation.image
      this.frameRate = animation.frameRate || 1
      this.frameBuffer = animation.frameBuffer || 2
    }
  }

  checkEnemiesCollision() {
    for (let i = 0; i < this.enemies.length; i++) {
      const enemy = this.enemies[i]
      if (enemy.shouldDraw) {
        if (
          this.position.x <= enemy.position.x + enemy.dimensions.width &&
          this.position.x + this.dimensions.width >= enemy.position.x &&
          enemy.position.y - (this.position.y + this.dimensions.height) <=
          0.5 &&
          enemy.position.y - (this.position.y + this.dimensions.height) > -0.3
        ) {
          enemy.destroyEnemy()
          return
        }

        if (
          this.position.x <= enemy.position.x + enemy.dimensions.width &&
          this.position.x + this.dimensions.width >= enemy.position.x &&
          this.position.y + this.dimensions.height > enemy.position.y &&
          this.position.y + this.dimensions.height <=
          enemy.position.y + enemy.dimensions.height
        ) {
          this.gameOver()
          break
        }
      }
    }
  }

  checkCoinCollision() {
    for (let i = 0; i < this.coins.length; i++) {
      const coin = this.coins[i]
      if (this.checkCollision(coin)) {
        if (coin.shouldDraw) {
          coin.getCoin()
          break
        }
      }
    }
  }

  checkHorizontalCollision() {
    for (let i = 0; i < this.collisionBlocks.length; i++) {
      const collisionBlock = this.collisionBlocks[i]
      if (this.checkCollision(collisionBlock)) {
        if (this.velocity.x < 0) {
          this.position.x =
            collisionBlock.position.x + collisionBlock.dimensions.width + 0.03
          break
        }
        if (this.velocity.x > 0) {
          this.position.x =
            collisionBlock.position.x - this.dimensions.width - 0.03
          break
        }
      }
    }
  }

  checkVerticalCollision() {
    for (let i = 0; i < this.collisionBlocks.length; i++) {
      const collisionBlock = this.collisionBlocks[i]
      if (this.checkCollision(collisionBlock)) {
        if (this.velocity.y < 0) {
          this.velocity.y = 0
          this.position.y = collisionBlock.position.y + collisionBlock.dimensions.height + 0.03
          break
        }
        if (this.velocity.y > 0) {
          this.velocity.y = 0
          this.position.y =
            collisionBlock.position.y - this.dimensions.height - 0.03
          break
        }
      }
    }
  }

  checkCollision(collisionBlock: CollisionBlock | Coin) {
    return (
      this.position.x <=
      collisionBlock.position.x + collisionBlock.dimensions.width &&
      this.position.x + this.dimensions.width >= collisionBlock.position.x &&
      this.position.y + this.dimensions.height >= collisionBlock.position.y &&
      this.position.y <=
      collisionBlock.position.y + collisionBlock.dimensions.height
    )
  }

  draw() {
    super.draw()
  }
}
