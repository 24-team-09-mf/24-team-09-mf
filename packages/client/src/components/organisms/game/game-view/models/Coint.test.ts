import { Coin } from '@/components/organisms/game/game-view/models/Coin'

import {
  HEIGHT_VIEW,
  WIDTH_VIEW,
  BLOCK_SIZE,
} from '@/components/organisms/game/game.constants'

describe('Тест класса Coin', () => {
  let canvas, ctx: CanvasRenderingContext2D, coin: Coin
  beforeEach(() => {
    canvas = document.createElement('canvas')
    canvas.width = WIDTH_VIEW
    canvas.height = HEIGHT_VIEW
    ctx = canvas.getContext('2d') as CanvasRenderingContext2D
    coin = new Coin({
      model: ctx,
      position: { x: 6 * BLOCK_SIZE, y: 3 * BLOCK_SIZE - 10 },
      frameRate: 4,
      frameBuffer: 15,
      imageSrc: '/assets/sprites/coin/coin.png',
      animations: {
        getCoin: {
          frameRate: 3,
          frameBuffer: 15,
          imageSrc: '/assets/sprites/coin/coinEffect.png',
        },
      },
    })
  })

  it('Проверяю инстанс класса', () => {
    expect(coin).toBeInstanceOf(Coin)
  })

  it('Проверяю изменение анимации', () => {
    const switchSpriteSpy = jest.spyOn(coin, 'switchSprite')
    coin.getCoin()
    expect(switchSpriteSpy).toHaveBeenCalledWith('getCoin')
  })

  it('Проверяю изменение анимации', () => {
    jest.useFakeTimers()
    coin.compeleAnimation = true
    coin.shouldDraw = true
    coin.getCoin()
    jest.runAllTimers()
    expect(coin.shouldDraw).toBe(false)
  })

  it('Проверяю вызов отрисовки', () => {
    const superDrawSpy = jest.spyOn(Coin.prototype, 'draw')
    coin.shouldDraw = true
    coin.compeleAnimation = false
    coin.draw()
    expect(superDrawSpy).toHaveBeenCalled()
  })

  it('Проверяю метод draw', () => {
    coin.draw()
    const events = ctx.__getEvents()
    expect(events).toMatchSnapshot()
  })
})
