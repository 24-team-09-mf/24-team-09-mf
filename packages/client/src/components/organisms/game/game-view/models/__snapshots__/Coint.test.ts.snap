// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тест класса Coin Проверяю метод draw 1`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 192,
      "dy": 86,
      "img": <img
        src="/assets/sprites/coin/coin.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
